tests/src/python/test_qgsvirtuallayertask.py

```diff
@@ -0,0 +1,72 @@
+# -*- coding: utf-8 -*-
+"""QGIS Unit tests for QgsVirtualLayerTask.
+
+.. note:: This program is free software; you can redistribute it and/or modify
+it under the terms of the GNU General Public License as published by
+the Free Software Foundation; either version 2 of the License, or
+(at your option) any later version.
+"""
+__author__ = 'Paul Blottiere'
+__date__ = '28/02/2018'
+__copyright__ = 'Copyright 2018, The QGIS Project'
+# This will get replaced with a git SHA1 when you do a git archive
+__revision__ = '$Format:%H$'
+
+import qgis  # NOQA
+import os
+
+from qgis.core import (
+    QgsProject,
+    QgsVectorLayer,
+    QgsApplication,
+    QgsVirtualLayerDefinition,
+    QgsVirtualLayerTask
+)
+from qgis.PyQt.QtCore import QCoreApplication
+from qgis.testing import start_app, unittest
+from utilities import unitTestDataPath
+
+start_app()
+
+
+class TestQgsVirtualLayerTask(unittest.TestCase):
+
+    def setUp(self):
+        self.testDataDir = unitTestDataPath()
+        self.success = False
+        self.fail = False
+
+    def onSuccess(self):
+        self.success = True
+
+    def onFail(self):
+        self.fail = True
+
+    def test(self):
+        l1 = QgsVectorLayer(os.path.join(self.testDataDir, "france_parts.shp"), "françéà", "ogr", QgsVectorLayer.LayerOptions(False))
+        self.assertEqual(l1.isValid(), True)
+        QgsProject.instance().addMapLayer(l1)
+
+        df = QgsVirtualLayerDefinition()
+        df.setQuery('select * from "françéà"')
+        task = QgsVirtualLayerTask(df)
+
+        ids = [f.id() for f in task.layer().getFeatures()]
+        self.assertEqual(len(ids), 0)
+
+        task.taskCompleted.connect(self.onSuccess)
+        task.taskTerminated.connect(self.onFail)
+
+        QgsApplication.taskManager().addTask(task)
+        while not self.success and not self.fail:
+            QCoreApplication.processEvents()
+
+        self.assertTrue(self.success)
+        self.assertFalse(self.fail)
+
+        ids = [f.id() for f in task.layer().getFeatures()]
```

Comment:
@pblottiere this test is a bit unstable - see https://dash.orfeo-toolbox.org/testDetails.php?test=67944698&build=312168

I think the id retrieval needs to be moved to onSuccess. Otherwise the task can be deleted by the manager before this line executes.